File: MechanicalSystem.m
classdef MechanicalSystem < handle
    properties  (Access = protected)
        x; % state vector
        x0; % initial value of the state vector
        st; % sampling periodi
        sigma_y=0; % noise standard deviation (if MIMO, it is a vector)
        t=0; % time
        umax=5; % maximum input
        u
        u0

        order=2; % system order
        num_output=1; % number of system outputs
        num_input=1; % number of system inputs
        scenario=1;

        output_names;
        input_names;
    end

    methods  (Access = public)

        % obj=MechanicalSystem(st)
        % create a mechanical model with sampling period st
        function obj=MechanicalSystem(st)
            obj.st=st;
            obj.x=zeros(2,1);
            obj.x0=zeros(2,1);
            obj.order=2;
            obj.num_input=1;
            obj.num_output=1;
            obj.u0=zeros(obj.num_input,1);

            for idx=1:obj.num_output
                obj.output_names{idx}=sprintf('output_%d',idx);
            end
            for idx=1:obj.num_input
                obj.input_names{idx}=sprintf('input_%d',idx);
            end
        end

        % (re)initialize the system to initial conditions
        function obj=initialize(obj)
            obj.x=obj.x0;
            obj.u=obj.u0;
            obj.t=0;
        end

        function obj=setScenario(obj,scenario)
            obj.scenario=scenario;
        end

        function obj=writeActuatorValue(obj,u)
            obj.u=u;
        end

        function u=readActuatorValue(obj)
            u=obj.u;
        end

        function y=readSensorValue(obj)
            y=obj.outputFunction();
        end



        % solve derivative(x)=f(x,u) during sampling period
        function obj=simulate(obj)
            usat=obj.saturationControlAction(obj.u);
            n=10;
            dt=obj.st/n;
            for idx=1:n
                obj.odeSolver(usat,dt,dt*idx);
            end
        end


        % return system sampling period
        function st=getSamplingPeriod(obj)
            st=obj.st;
        end

        % return number of system outputs
        function num_output=getOutputNumber(obj)
            num_output=obj.num_output;
        end

        % return number of system inputs
        function num_input=getInputNumber(obj)
            num_input=obj.num_input;
        end

        function umax=getUMax(obj)
            umax=obj.umax;
        end

        function output_names=getOutputName(obj)
             output_names=obj.output_names;
        end

        function input_names=getInputName(obj)
            input_names=obj.input_names;
        end

        % show system
        function obj=show(obj)
            fprintf('This system has %d outputs:\n',obj.num_output)
            for io=1:obj.num_output
                fprintf('- %s\n',obj.output_names{io})
            end

            fprintf('This system has %d inputs:\n',obj.num_input)
            for ii=1:obj.num_input
                fprintf('- %s, with maximum limit = %f\n',obj.input_names{ii},obj.umax(ii))
            end
        end

    end

    methods  (Access = protected)

        % derivative(x)=f(x,u)
        function Dx=stateFunction(obj,x,u,t)
            Dx=0;
        end

        % ouput=g(x,u)
        function y=outputFunction(obj)
            y=obj.x+obj.sigma_y.*randn(length(obj.sigma_y),1);
        end

        % ODE solver (using RK4)
        function odeSolver(obj,u,st,t)
            % Runge Kutta 4
            k_1 = obj.stateFunction(obj.x,u,t);
            k_2 = obj.stateFunction(obj.x+0.5*st*k_1,u,t);
            k_3 = obj.stateFunction(obj.x+0.5*st*k_2,u,t);
            k_4 = obj.stateFunction(obj.x+k_3*st,u,t);
            obj.x=obj.x+(1/6)*(k_1+2*k_2+2*k_3+k_4)*st;
        end

        % saturate control action
        function usat=saturationControlAction(obj,u)
            for iu=1:obj.num_input
                if (u(iu)>obj.umax(iu))
                    usat(iu,1)=obj.umax(iu);
                elseif (u(iu)<-obj.umax(iu))
                    usat(iu,1)=-obj.umax(iu);
                else
                    usat(iu,1)=u(iu);
                end
            end
        end
    end
end

File: mechanical_system.py
import numpy as np

class MechanicalSystem:
    def __init__(self, st):
        """
        Initialize the mechanical system with a given sampling period.
        """
        self.st = st  # Sampling period
        self.x = np.zeros(2)  # State vector
        self.x0 = np.zeros(2)  # Initial state vector
        self.order = 2  # System order
        self.num_input = 1  # Number of system inputs
        self.num_output = 1  # Number of system outputs
        self.u0 = np.zeros(self.num_input)  # Initial input values
        self.u = self.u0.copy()  # Current input values
        self.sigma_y = 0  # Noise standard deviation
        self.t = 0  # Time variable
        self.umax = np.array([5])  # Maximum input value
        self.scenario = 1  # Scenario identifier
        
        # Define output and input names
        self.output_names = [f'output_{i+1}' for i in range(self.num_output)]
        self.input_names = [f'input_{i+1}' for i in range(self.num_input)]
    
    def initialize(self):
        """
        Reset the system to its initial conditions.
        """
        self.x = self.x0.copy()
        self.u = self.u0.copy()
        self.t = 0
    
    def set_scenario(self, scenario):
        """
        Set the current scenario for the system.
        """
        self.scenario = scenario
    
    def write_actuator_value(self, u):
        """
        Update the actuator input value.
        """
        self.u = np.array(u)
    
    def read_actuator_value(self):
        """
        Read the current actuator input value.
        """
        return self.u
    
    def read_sensor_value(self):
        """
        Read the sensor output value.
        """
        return self.output_function()
    
    def simulate(self):
        """
        Solve dx/dt = f(x, u) over the sampling period using numerical integration.
        """
        usat = self.saturation_control_action(self.u)  # Apply saturation control
        n = 10  # Number of integration steps
        dt = self.st / n  # Time step for integration
        for i in range(n):
            self.ode_solver(usat, dt, dt * (i + 1))
    
    def get_sampling_period(self):
        """
        Return the system's sampling period.
        """
        return self.st
    
    def get_output_number(self):
        """
        Return the number of system outputs.
        """
        return self.num_output
    
    def get_input_number(self):
        """
        Return the number of system inputs.
        """
        return self.num_input
    
    def get_umax(self):
        """
        Return the maximum allowed input value.
        """
        return self.umax
    
    def get_output_names(self):
        """
        Return the names of the system outputs.
        """
        return self.output_names
    
    def get_input_names(self):
        """
        Return the names of the system inputs.
        """
        return self.input_names
    
    def show(self):
        """
        Display information about the system's inputs and outputs.
        """
        print(f'This system has {self.num_output} outputs:')
        for name in self.output_names:
            print(f'- {name}')
        
        print(f'This system has {self.num_input} inputs:')
        for i, name in enumerate(self.input_names):
            print(f'- {name}, with maximum limit = {self.umax[i]}')
    
    def state_function(self, x, u, t):
        """
        Define the state transition function dx/dt = f(x, u).
        """
        return np.zeros_like(x)
    
    def output_function(self):
        """
        Define the system output function.
        """
        return self.x + self.sigma_y * np.random.randn(len(np.atleast_1d(self.sigma_y)))
    
    def ode_solver(self, u, st, t):
        """
        Solve the system's differential equation using the Runge-Kutta 4th Order method.
        """
        k1 = self.state_function(self.x, u, t)
        k2 = self.state_function(self.x + 0.5 * st * k1, u, t)
        k3 = self.state_function(self.x + 0.5 * st * k2, u, t)
        k4 = self.state_function(self.x + st * k3, u, t)
        self.x += (1/6) * (k1 + 2*k2 + 2*k3 + k4) * st
    
    def saturation_control_action(self, u):
        """
        Apply saturation limits to the control input.
        """
        usat = np.clip(u, -self.umax, self.umax)
        return usat
